Django recipe exercise for TK onboarding

clicktravel-andrew opened the pull request.

clicktravel-andrew committed:
Adding Ignore for PyCharm project files

clicktravel-andrew committed:
Adding project files

Adding traveix file for CI/CD
Adding flake8 config for linting exeptions
Adding docker setup files and initial Django app folder and files

clicktravel-andrew committed:
adding core app and wit_for_db

clicktravel-andrew committed:
Added model and model tests

clicktravel-andrew committed:
Prefer recipe.description to be a TextField

clicktravel-andrew committed:
Adding recipse view, serializer, tests and wiring together

clicktravel-andrew committed:
Adding postman script for getting recipes

clicktravel-andrew committed:
Add feature to create a recipe

clicktravel-andrew committed:
Adding create recipe script

clicktravel-andrew committed:
Updates for adding recipes and seeing a detailed view

clicktravel-andrew committed:
Adding filter by recipe name

clicktravel-andrew committed:
Removing filtering printline

clicktravel-andrew committed:
removing errant whitespace

clicktravel-andrew committed:
Adding test to cover deleting a recipe

clicktravel-andrew committed:
improving test, only need one receipe for delete testing

clicktravel-andrew committed:
Adding delete postman script

clicktravel-andrew committed:
Adding ability to PATCH recipes

volker-travelperk commented on app/core/models.py:
```diff
@@ -0,0 +1,23 @@
+
+class Recipe(models.Model):
+    """Recipe object"""
+    name = models.CharField(max_length=31)
```
31? Any specific reason? :sweat_smile:  (I'd not asked for 42)

clicktravel-andrew replied:
I wanted something small to force titles of recipes to be short, the description field can be used for full context. In the training course the fields are all 255 which is a bit too long, perhaps I've been a bit too mean with the default value set here.

^^ incidentally, I just checked and that comment is 254 chars long 🤣

volker-travelperk commented on app/recipe/tests/test_recipe_api.py:
```diff
@@ -0,0 +1,172 @@
+    return reverse('recipe:recipe-detail', args=[recipe_id])
+
+
+def sample_ingredient(recipe, name='love'):
```
out of the scope of this exercise, but fyi, we use this library to generate dummy data:
https://factoryboy.readthedocs.io/en/latest/orms.html#django

clicktravel-andrew replied:
👍  this is good to know

volker-travelperk commented on app/recipe/tests/test_recipe_api.py:
```diff
@@ -0,0 +1,172 @@
+        serializer = RecipeSerializer(recipes, many=True)
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.data, serializer.data)
```
Better test against the raw data. The serializer is part of the API and you don't test it like that

clicktravel-andrew replied:
👍  I'll have a look at improving this test

volker-travelperk commented on app/recipe/tests/test_recipe_api.py:
```diff
@@ -0,0 +1,172 @@
+        recipe = Recipe.objects.get(id=response.data['id'])
+        ingredients = recipe.ingredients.all()
+        self.assertEqual(ingredients.count(), 3)
+        self.assertEqual(len(ingredients.filter(name=ingredient1['name'])), 1)
```
#premature_improvement (but so you are aware)
```suggestion
        self.assertEqual(ingredients.filter(name=ingredient1['name']).count(), 1)
```
Like this you issue a `COUNT` to the database, otherwise you'd retrieve the full object which in extreme cases would be slower

clicktravel-andrew replied:
Thanks, I'll make the change as I'm certain to be referring to this project again in future.

volker-travelperk commented on requirements.txt:
```diff
@@ -0,0 +1,5 @@
+Django>=2.1.3,<2.2.0
```
Better use fixed versions in your "production environment" to have a determined version deployed. Ranges are better suited for libraries to be as compatible as possible.

clicktravel-andrew replied:
I think I'll fix this at 2.2.24 in the first instance as 2.2 is the LTS versions which will support until 04/2022 and hopefully there will be less chance of breaking changes compares to 3.x. Ideally I'd have gone for the same version used in the TK repos

clicktravel-andrew committed:
Setting DJanog to a fixed version following PR comments

clicktravel-andrew committed:
Updates following PR review

This update:
-  removes the use of the Serializer in the test
- Changes the meachanism by which items in the database are counted.
